### ISSUE TITLE
r/aws_kinesis_firehose_delivery_stream: Delivery to Amazon Elasticsearch Service domain in VPC

### ISSUE THREAD

**Issue by ewbankkit** [terraform-providers/terraform-provider-aws#13015]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon Kinesis Data Firehose can now deliver streaming data to an Amazon Elasticsearch Service domain in an Amazon VPC. 

### New or Affected Resource(s)

* [`aws_kinesis_firehose_delivery_stream`](https://www.terraform.io/docs/providers/aws/r/kinesis_firehose_delivery_stream.html)

### Potential Terraform Configuration

```hcl
resource "aws_kinesis_firehose_delivery_stream" "example" {
  elasticsearch_configuration {
    vpc_configuration {
      subnet_ids = ["..."]
      security_group_ids = ["..."]
      role_arn = "..."
    }
  }
}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2020/04/amazon-kinesis-data-firehose-adds-support-for-streaming-data-delivery-to-amazon-elasticsearch-service-in-amazon-vpc/).

Requires [AWS SDK v1.30.13](https://github.com/aws/aws-sdk-go/releases/tag/v1.30.13):
* https://github.com/terraform-providers/terraform-provider-aws/pull/12970

**Comment by martin-paulus** [comment]
Consistentcy note: The `aws_elasticsearch_domain` resource uses a `vpc_options` block. The `aws_lambda_function` resource uses a `vpc_config` block.

Please consider one of these two options to prevent introducing a third label. Hopefully, in time, we can use only one label for this type of block (in AWS resources).

**Comment by martin-paulus** [comment]
Prerequisite https://github.com/terraform-providers/terraform-provider-aws/pull/12970 has been merged, and then released in [v2.60.0](https://github.com/terraform-providers/terraform-provider-aws/blob/v2.60.0/CHANGELOG.md) of the AWS provider.

**Comment by rajholla** [comment]
I just started working on this. I will link the PR once I am done.

**Comment by ddiawara** [comment]
still have the same error even if update provider 
```yaml
provider "aws" {
  version = "~> 2.60.0" # "~> 2.32"
  alias   = "ireland"
  region  = "eu-west-1"
  assume_role {
    role_arn = "arn:aws:iam::${var.aws_account}:role/ts-terraformRole"
  }
}
```

**Comment by martin-paulus** [comment]
@ddiawara Sure, because "aws" provider version 2.60.0 contains support for the AWS SDK version that supports VPC configuration. But the Terraform support for VPC configuration is present in the PR that @rajholla prepared. That PR is still unmerged.

**Comment by ddiawara** [comment]
is it possible to use it now while waiting for the merge to be accepted.
I have updated the provider 2.60.0 version and I have implemented the Aws-sdk version but I keep getting the same error message.

**Comment by martin-paulus** [comment]
@ddiawara In order to use (test) @rajholla 's contribution before it's merged, I believe you should follow this guide: https://www.terraform.io/docs/extend/writing-custom-providers.html

First, fetch the sources from: https://github.com/rajholla/terraform-provider-aws/tree/firehose-es-vpc-support
Ensure you checkout the **firehose-es-vpc-support** branch.

Then build the Terraform provider using the **Writing Custom Providers** guide: https://www.terraform.io/docs/extend/writing-custom-providers.html#building-the-plugin

Note: The essential changes are in: https://github.com/rajholla/terraform-provider-aws/blob/firehose-es-vpc-support/aws/resource_aws_kinesis_firehose_delivery_stream.go
But I am not sure whether you can stop after building this, or whether you should build the entire **aws** provider.

After a successful build, don't forget to install the "custom provider": https://www.terraform.io/docs/configuration/providers.html#third-party-plugins

And finally, you should configure your Terraform configuration to select your custom provider.

**Comment by martin-paulus** [comment]
Thank you @rajholla , I just now successfully deployed a Kinesis Firehose delivery stream with VPC access!

**Comment by ddiawara** [comment]
> Thank you @rajholla , I just now successfully deployed a Kinesis Firehose delivery stream with VPC access!

Yo yeah what do you do because it doesn't work for me 
this is what is do : 
```sh
mkdir -p $HOME/development/terraform-providers/; cd $HOME/development/terraform-providers/
git clone https://github.com/rajholla/terraform-provider-aws.git
cd terraform-provider-aws && git checkout firehose-es-vpc-support
make tools
make build
cp $GOPATH/bin/terraform-provider-aws ~/.terraform.d/plugins
# in a terraform code
![image](https://user-images.githubusercontent.com/62718811/82463356-0618b380-9abd-11ea-8352-a77e75505ebc.png)

![image](https://user-images.githubusercontent.com/62718811/82463385-0f098500-9abd-11ea-82a6-bb6534842647.png)

terraform init

![image](https://user-images.githubusercontent.com/62718811/82463476-2e081700-9abd-11ea-8831-7e4e5a949775.png)
terraform apply 
![image](https://user-images.githubusercontent.com/62718811/82463548-3fe9ba00-9abd-11ea-8d6e-9f5831e8c260.png)

i don't know what i forget

**Comment by martin-paulus** [comment]
@ddiawara Please try renaming your provider binary with a filename that includes a version. Also, consider configuring Terraform to use *aws* as the provider name for your custom binary.

This is what I did:

```
git clone git@github.com:rajholla/terraform-provider-aws.git
cd terraform-provider-aws/
git checkout firehose-es-vpc-support

docker run -it --rm -v "$PWD":/usr/local/src/tf-provider -w /usr/local/src/tf-provider golang go build -o terraform-provider-awsCustomMain -v

mkdir ~/.terraform.d/plugins
cp terraform-provider-awsCustomMain ~/.terraform.d/plugins/terraform-provider-aws_v2.63.0
```

Note: Here  I used a Docker container to compile the sources, if you like you could instead use a native golang compiler which you installed.

Select the custom build **aws** provider in Terraform:

```
terraform {
  required_providers {
    aws = "2.63.0"
  }
}
```

Finally, run:

```
terraform init
terraform apply
```

**Comment by martin-paulus** [comment]
@ddiawara Your VPC config block is misnamed, it should look something like:

```
resource "aws_kinesis_firehose_delivery_stream" "es-delivery" {
  elasticsearch_configuration {
    vpc_config {
      role_arn           = aws_iam_role.firehose-vpc.arn
      security_group_ids = [aws_security_group.firehose.id]
      subnet_ids         = slice(local.vpc.private_subnets, 0, var.es_instance_count)
    }
  }
}
```

So use _vpc_config_ instead of _vpc_configuration_.

**Comment by ddiawara** [comment]
I thought I was going crazy, haahhaha thank you for your excellent work.

**Comment by papovyr** [comment]
Hi,
Any timeline when `vpc_config` will be released for `elasticsearch_configuration` ?

**Comment by Kevc1** [comment]
Is vpc_config in any currently available terraform version so I don't need to compile my own version ?
Thanks

**Comment by monolithh** [comment]
Do we know when VPC block will be available for ES configuration?

**Comment by bflad** [comment]
Support for this functionality has been merged and will release with version 3.5.0 of the Terraform AWS Provider, later this week. Thanks to @rajholla for the implementation. 👍
